### PR TITLE
Initial version with Django and Vue

### PR DESCRIPTION
**Pull request by JokeWaumans**

### CODE REVIEW

File: py-requirements.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+setuptools==58.2.0
+neomodel==5.1.2
+django-neomodel==0.1.1
+ruamel.yaml==0.18.2
+# pyyaml>=6.0 #5.4.1
+# docker-compose==1.29.2
+django>=4.0,<5.0 
+djangorestframework==3.14.0
+exrex==0.11.0
```
Comment (JasperCraeghs) [inline py-requirements.txt]: Is `exrex` used?

Comment (JokeWaumans) [reply py-requirements.txt]: No, not anymore, I will remove that.

Comment (JokeWaumans) [reply py-requirements.txt]: Resolved in a55bc338be31f243b75f6ab42ebe5998d907a81d

File: traceabilityViewer/myapp/scripts/create_database.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+with open("../config.yml", "r", encoding="utf-8") as open_file:
+    configuration = YAML().load(open_file)
+
+groups_list = list(configuration["layers"]) + list(configuration["layers"].values())
+unique_groups = list(dict.fromkeys(groups_list))
+
+def validate_keyword(configuration, keyword, expected_type):
+    if configuration.get(keyword) is None:
+        raise TypeError(f"Expected {keyword} to be in the configuration file; got 'None'")
```
Comment (JasperCraeghs) [inline traceabilityViewer/myapp/scripts/create_database.py]:
```
if  keyword not in configuration:
    raise ValueError(f"Failed to find mandatory parameter {keyword!r} in the configuration file")

Comment (JokeWaumans) [reply traceabilityViewer/myapp/scripts/create_database.py]: Resolved in 8dd5a7983cdeadf33ba43e73eb1146e3d6e9d14f

File: traceabilityViewer/myapp/static/components/alert.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    props: {
+      alertData: {
+        type: Object,
+      },
+    },
+    emits: ["closeAlert"],
+    setup(props, { emit }) {
+        let show = Vue.ref(false);
+        let alert = Vue.ref(null);
+        let alertData = Vue.toRef(props, "alertData");
```
Comment (JasperCraeghs) [inline traceabilityViewer/myapp/static/components/alert.js]: Why do you make the prop `alertData` reactive? Is its value supposed to change?

File: traceabilityViewer/myapp/static/components/traceabilityViewer.js [hunk 4]
```diff
@@ -124,21 +113,20 @@ app.component("traceability-viewer", {
...
-    function onAlert(data) {
-      console.log(data)
-      errorText.value = data.message;
-      info.value.innerText = data.title;
-      alert.value.scrollIntoView({behavior: "smooth"})
-      // alert.value.style.display = "block";
+    function onAlert(alertData) {
+      alertData["number"] = alertNumber;
+      alerts.value.push(alertData);
+      alertNumber += 1;
```
Comment (JasperCraeghs) [inline traceabilityViewer/myapp/static/components/traceabilityViewer.js]: Instead of keeping track of the number to use for the next alert in this variable `alertNumber`, you could extract the number of the last element in `alerts` and add 1. Would this be cleaner?

File: traceabilityViewer/myapp/static/components/alert.js (hunk 3 above)
Comment (JokeWaumans) [reply traceabilityViewer/myapp/static/components/alert.js]: No, indeed, it doesn't need to be reactive.

File: traceabilityViewer/myapp/static/components/traceabilityViewer.js [hunk 5]
```diff
@@ -124,21 +113,20 @@ app.component("traceability-viewer", {
...
     }
 
-    // function close() {
-    //   alert.value.style.display = "none";
-    //   error.value.className = "{show: false}"
-    //   document.getElementById("arrow-button").innerHTML = "&#8964";
-    //   info.value.innerText = "";
-    //   show = false;
-    // }
+    function closeAlert(number){
```
Comment (JasperCraeghs) [inline traceabilityViewer/myapp/static/components/traceabilityViewer.js]: `removeAlert` would be a more suitable name as that's essentially what this function does.

If you pass the entire `alert` object as argument, you can simply remove the element with `array.splice` instead of creating a second array with `array.filter` with the element with the given number filtered out.
 https://stackoverflow.com/a/50808869/10539205

File: traceabilityViewer/myapp/static/components/alert.js (hunk 3 above)
Comment (JokeWaumans) [reply traceabilityViewer/myapp/static/components/alert.js]: Resolved in bbec26191854a98b90bf328174a413956bc69c4d

File: traceabilityViewer/myapp/static/components/traceabilityViewer.js (hunk 4 above)
Comment (JasperCraeghs) [reply traceabilityViewer/myapp/static/components/traceabilityViewer.js]: No, please use this functionality of `v-for`: https://vuejs.org/guide/essentials/list#maintaining-state-with-key

File: traceabilityViewer/myapp/static/components/traceabilityViewer.js (hunk 5 above)
Comment (JokeWaumans) [reply traceabilityViewer/myapp/static/components/traceabilityViewer.js]: The name of the function is changed in commit cc7c1c8f2861d23b0c507a6645f5137270d0e5e7

File: traceabilityViewer/myapp/static/components/traceabilityViewer.js (hunk 4 above)
Comment (JokeWaumans) [reply traceabilityViewer/myapp/static/components/traceabilityViewer.js]: Doesn't need the `key` to be unique? The errors are not unique, so to remove an alert I need a unique key such as a number.

Comment (JasperCraeghs) [reply traceabilityViewer/myapp/static/components/traceabilityViewer.js]: I see. The backend should provide a unique identifier. It could include the `id()` of the caught `Exception` instance and pass this as additional information.

Comment (JokeWaumans) [reply traceabilityViewer/myapp/static/components/traceabilityViewer.js]: I have committed a suggestion af950240f1877f31961b7ebf219091ec94189253

File: traceabilityViewer/myapp/views.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+        except CypherSyntaxError as error:
+            return Response(status=status.HTTP_400_BAD_REQUEST, data=f"An error occured in function {func.__name__}.\n{error}")
+        except (BufferError, TypeError, ValueError) as error:
+            LOGGER.error(repr(error))
+            return Response(status=status.HTTP_400_BAD_REQUEST,
+                            data=f"An {type(error)} occured in function {func.__name__}.\n" + traceback.format_exc()
+                            )
+        except Exception as error:
+            return Response(status=status.HTTP_400_BAD_REQUEST,
+                            data=f"Something went wrong in function {func.__name__}.\n{error}")
```
Comment (JasperCraeghs) [inline traceabilityViewer/myapp/views.py]:
```suggestion
        except Exception as error:  # pylint: disable=broad-exception-caught
            LOGGER.error(repr(error))
            error_status = status.HTTP_400_BAD_REQUEST
            data = {
                "identifier": id(error),
                "message": f"An error occurred in function {func.__name__}:",
            }
            if isinstance(error, (BufferError, TypeError, ValueError)):
                data["message"] += f"\n{traceback.format_exc()}"
            else:
                data["message"] += f"\n{error}"
            return Response(status=error_status, data=data)
```

File: traceabilityViewer/myapp/views.py [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+                            f"Expected Node or Relationship types to be returned from the query; "
+                            f"got {type(element)}"
+                        )
+                        return Response(data=error)
+
+            else:
+                error = TypeError(
+                    f"Expected Node or Relationship types to be returned from the query; " f"got {type(element)}"
+                )
+                return Response(data=error)
```
Comment (JasperCraeghs) [inline traceabilityViewer/myapp/views.py]: This function `searchConnectedNodes` shares a lot of code with `query`. This should be refactored immediately to avoid double work, like the integration of the new decorator `error_handling` properly, which you did well in `query` but not here.

Comment (JokeWaumans) [reply traceabilityViewer/myapp/views.py]: Resolved in 47c56989554873cc575657b8504375f31c8246f5

File: config.yml [hunk 8]
```diff
@@ -1,4 +1,4 @@
-traceability_export: build/doc/${PRODUCT}/traceability_export.json
+traceability_export: traceability_export.json
```
Comment (JasperCraeghs) [inline config.yml]: `./traceability_export.json` clarifies that a relative path is supported. Is this configuration value necessary? It's also defined as an environment variable in https://github.com/melexis/traceability-viewer/pull/2/commits/a3a2399ef8198823868b898ad42d20bd6660e0c4#diff-a3046da0d15a27e89f2afe639b25748a7ad4d9290af3e7b1b6c1a5533c8f0a8cR1

Comment (JokeWaumans) [reply config.yml]: I think I can remove that part because it is copied in the docker container.

Comment (JokeWaumans) [reply config.yml]: Resolved in 3be821469f504e71946bde188d79887341ad7b1c

File: README.md [hunk 9]
```diff
@@ -1 +1,297 @@
...
+When you use the [traceability-plugin](https://melexis.github.io/sphinx-traceability-extension/index.html), you can [export](https://melexis.github.io/sphinx-traceability-extension/configuration.html#export) the documentation items as a JSON.
+
+### html_documentation_root
+
+> [!NOTE]
+> `html_documentation_root` is not a required configuration parameter.
+
+The html root to the documentation of that item. This can also be the path to a local html folder.
+
+This is only workable when you use the exported JSON of the [traceability-plugin](https://melexis.github.io/sphinx-traceability-extension/configuration.html#export-to-json).
```
Comment (JasperCraeghs) [inline README.md]: What do you mean by this sentance?

File: README.md [hunk 10]
```diff
@@ -1 +1,297 @@
...
+When hovering a nodes, the name of that node is shown.
+In the previous image, the black node at the bottom was hovered by the mouse.
+
+#### HTML link in Info of Node
+
+When `html_documentation_root` is specified, the name of the node will be a link to the documentation of that specific node.
+
+#### Highlight Connected Nodes
+
+When the button with the eye is clicked, the nodes connecting to the selected node will be highlighted. The rest of the nodes get a lower opacity.
```
Comment (JasperCraeghs) [inline README.md]: "nodes connecting to the selected node"

How far? Are only directly connected nodes highlighted?

```suggestion
The button with the eye-icon highlights all nodes that are directly connected to the selected node and makes other nodes transparent.
```

File: README.md [hunk 11]
```diff
@@ -1 +1,297 @@
...
+![hide group](images/hide_link.png)
+
+
+#### Selecting a Node
+
+When clicking a node, this node will be selected and the info will be visible in the top left corner.
+
+![select node](images/select_node.png)
+
+The image shows the info of the selected node. There are two more buttons showed as well. These two buttons pop up when a node is selected. These will be explained in [highlight connecting nodes](####highlight_connecting_nodes) and [Add/show connecting nodes](####add/show_connecting_nodes).
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
You can click on a node to select it. Additional info about the node will be displayed in the top left corner and
two additional buttons pop up:

- [Highlight Connecting Nodes](####Highlight-Connected-Nodes)
- [Add/Show Connecting Nodes](####Add/Show-Connecting-Nodes)

![select node](images/select_node.png)
```
Spaces must be replaced by `-` when referencing titles.

File: README.md [hunk 12]
```diff
@@ -1 +1,297 @@
-# traceability-viewer
\ No newline at end of file
+# traceability-viewer
```
Comment (JasperCraeghs) [inline README.md]: Please delete this top-level title and change the level of all others so that they are rendered larger.

File: README.md [hunk 13]
```diff
@@ -1 +1,297 @@
...
+  - [html_documentation_root](#html_documentation_root)
+  - [layered](#layered)
+  - [layers](#layers)
+  - [item_colors](#item_colors)
+  - [link_colors](#link_colors)
+  - [backwards_relationships](#backwards_relationships)
+  - [visualised_properties](#visualised_properties)
+- [Usage](#Usage)
+
+
```
Comment (JasperCraeghs) [inline README.md:44]: Please add an "Overview" chapter that explains what the traceability viewer can be used for, what it does (it visualizes a collection of traceable items in a network graph), and that it uses the JSON database of our traceability plugin as input.

This should give the reader an idea whether this tool could be useful to them or not.

File: README.md [hunk 14]
```diff
@@ -1 +1,297 @@
...
+# copy traceability_export.json file to project folder
+cp /path/to/traceability_export.json .
+# copy example .env to your .env
+cp .env.example .env
+```
+Lastly, use docker compose to start the application.
+
+```
+docker compose up --build --remove-orphans
+```
```
Comment (JasperCraeghs) [inline README.md:65]: Please explain why we chose to make use of the Docker Compose plugin and that the `.env` file is meant to store and define environment variables and requires modification.

How can the web application and database be accessed after starting the application?

File: README.md [hunk 15]
```diff
@@ -1 +1,297 @@
...
+                "<target-item-name-4>",
+                ...
+            ],
+        }
+    },
+    ...
+]
+```
+
+When you use the [traceability-plugin](https://melexis.github.io/sphinx-traceability-extension/index.html), you can [export](https://melexis.github.io/sphinx-traceability-extension/configuration.html#export) the documentation items as a JSON.
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
Our Sphinx plugin [mlx.traceability](https://melexis.github.io/sphinx-traceability-extension/index.html) can be configured to
[export](https://melexis.github.io/sphinx-traceability-extension/configuration.html#export) its collection of traceable documentation items in JSON format.
```

File: README.md [hunk 16]
```diff
@@ -1 +1,297 @@
...
+
+```
+docker compose up --build --remove-orphans
+```
+
+## Configuration
+
+The configuration takes the form of a YAML file. The [config.yml](config.yml) is an example that requires modification.
+
+### traceability_export
```
Comment (JasperCraeghs) [inline README.md]: This is not a configuration parameter. Please move this to section "getting started".

File: README.md [hunk 17]
```diff
@@ -1 +1,297 @@
...
+    ...
+]
+```
+
+When you use the [traceability-plugin](https://melexis.github.io/sphinx-traceability-extension/index.html), you can [export](https://melexis.github.io/sphinx-traceability-extension/configuration.html#export) the documentation items as a JSON.
+
+### html_documentation_root
+
+> [!NOTE]
+> `html_documentation_root` is not a required configuration parameter.
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
> `html_documentation_root` is an optional configuration parameter.
```
Please adjust this everywhere.

File: README.md [hunk 18]
```diff
@@ -1 +1,297 @@
...
+
+The html root to the documentation of that item. This can also be the path to a local html folder.
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
The root of the the HTML project documentation; either a pathname or a URL.
```

File: README.md [hunk 19]
```diff
@@ -1 +1,297 @@
...
+
+#### Zoom buttons
+
+There are three buttons for zooming, respectively zoom in, zoom out and zoom till the network fits the canvas.
+
+![zoom buttons](images/zoom.png)
+
+#### Errors
+
+When an error occurs you get a notification at the top of the page as shown in following example where an invalid Cypher query is submitted:
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
When an error occurs, a notification pops up at the top of the page. In the example below, an invalid Cypher query was submitted:
```

File: README.md [hunk 20]
```diff
@@ -1 +1,297 @@
...
+
+![error](images/error.png)
+
+You can show the error by clicking **Show error message**:
+
+![error](images/show_error.png)
+
+You can close the error by clicking **X**.
+If the error is not closed and another error occurs, all errors will be listed at the top of the page.
+You will be redirected to the hight of the newest error.
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
If the error is not closed and another error occurs, the new error pops up at the top.
```

File: README.md [hunk 21]
```diff
@@ -1 +1,297 @@
...
+
+## Bugs
+
+1. When showing the labels/names of the nodes, there is allways one node that turns black.
+1. When using the input fields, the autocompletion list will always look at the last part of the sentence, so when adding something in the middle the autocompletion will stay the same (focused on the last part).
+
+
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
1. When showing the labels/names of the nodes, there is always one node that turns black.
2. When using the input fields, the autocompletion list will always look at the last part of the sentence. So, when adding something in the middle, the autocompletion does not adapt.
```

File: README.md [hunk 22]
```diff
@@ -1 +1,297 @@
...
+
+## Features
+
+- Errors could be popping up, you need to close it before you can go further. That way the user can't miss it.
```
Comment (JasperCraeghs) [inline README.md:319]: Do you mean that certain errors should become blocking?

File: README.md [hunk 23]
```diff
@@ -1 +1,297 @@
...
+You will be redirected to the hight of the newest error.
+
+## Bugs
+
+1. When showing the labels/names of the nodes, there is allways one node that turns black.
+1. When using the input fields, the autocompletion list will always look at the last part of the sentence, so when adding something in the middle the autocompletion will stay the same (focused on the last part).
+
+
+
+## Features
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
## Planned Features
```

File: README.md [hunk 24]
```diff
@@ -1 +1,297 @@
...
+
+- Errors could be popping up, you need to close it before you can go further. That way the user can't miss it.
+- Faster rendering; open to suggestions
+- Instead of disabling the button when submitted, stop the previous request when another submit is done.
+- A node that needs a layer but the group is not specified, better to place it on the same hight as the first node that is connected to it.
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
- A node that needs a layer but the group is not specified, better to place it on the same height
  as the first node that is connected to it.
```
Please rephrase this sentence.

File: README.md [hunk 25]
```diff
@@ -1 +1,297 @@
...
+- Reset button when in a filter.
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
- Add a reset button to remove/undo the active filter.
```

File: py-requirements.txt [hunk 26]
```diff
@@ -0,0 +1,8 @@
+setuptools==58.2.0
+neomodel==5.1.2
+django-neomodel==0.1.1
+ruamel.yaml==0.18.2
+# docker-compose==1.29.2
```
Comment (JasperCraeghs) [inline py-requirements.txt]:
```suggestion
```

File: traceabilityViewer/app/views.py [hunk 27]
```diff
@@ -0,0 +1,313 @@
...
+                        for node_name in [link.source, link.target]:
+                            node = DocumentItem.nodes.get(name=node_name)
+                            node = node.node_data
+                            nodes.add(node)
+                    else:
+                        raise TypeError(
+                            f"Expected Node or Relationship type to be returned from the query; "
+                            f"got {type(element)}"
+                        )
+
```
Comment (JasperCraeghs) [inline traceabilityViewer/app/views.py]:
```suggestion
```

File: traceabilityViewer/traceabilityViewer/settings.py [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+
+# Build paths inside the project like this: BASE_DIR / 'subdir'.
+BASE_DIR = Path(__file__).resolve().parent.parent
+
+
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/4.2/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = "django-insecure-9@1+y@2v04x9t4r=ih5-%j7ueh(84j6($3tt6js&pd%ozl91du"
```
Comment (JasperCraeghs) [inline traceabilityViewer/traceabilityViewer/settings.py]: What's this for?

**Review by JasperCraeghs — CHANGES_REQUESTED** [review]

File: README.md (hunk 9 above)
Comment (JokeWaumans) [reply README.md]: I think it is only useful when you use the traceability plugin, but actually it can be any documentation that is composed of attributes of that element

Comment (JasperCraeghs) [reply README.md]: The JSON database must contain the path to the RST document and the item ID/name must be appended to the HTML filename with a `#`. The traceability plugin does both.

```suggestion
This configuration parameter can only be used if [the JSON database](###traceability_export) contains a ``"document"`` attribute with the relative path to the reST file in which the item is defined,
and if the item ID/name is appended to the HTML filename with a `#`. 
Our [traceability-plugin](https://melexis.github.io/sphinx-traceability-extension/configuration.html#export-to-json) does both.
```

File: README.md [hunk 29]
```diff
@@ -1 +1,300 @@
...
+```
+visualised_properties:
+    - name
+    - document
+```
+
+If [`html_documentation_root`](#html_documentation_root) and `visualised_properties` are defined but `document` is not in the list, it automatically adds `document` to the end of the list.
+
+If `visualised_properties` is defined but `name` is missing, it automatically adds `name` to the end of the list
+
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
> `visualised_properties` is an optional configuration parameter.

A list of all item properties to display after selecting a node. By default, its name/ID will be displayed, and,
in case [`html_documentation_root`](#html_documentation_root) has been configured,
the link to the item definition in your project documentation,.
```

File: traceabilityViewer/traceabilityViewer/settings.py (hunk 28 above)
Comment (JokeWaumans) [reply traceabilityViewer/traceabilityViewer/settings.py]: Django will refuse to start if `SECRET_KEY` is not set.
https://docs.djangoproject.com/en/5.0/ref/settings/#std:setting-SECRET_KEY

Comment (Letme) [reply traceabilityViewer/traceabilityViewer/settings.py]: it is needed. Should end up in .env file and here you should use `python-decouple` with `config` call.

File: README.md (hunk 9 above)
Comment (JokeWaumans) [reply README.md]: Suggestion applied in a7001e19f8f006e609f1a4d6a711a01590a42923

File: README.md (hunk 24 above)
Comment (JokeWaumans) [reply README.md]: Resolved in 3bc1af842099640e02bb7e825ab4b318196169b1

File: traceabilityViewer/traceabilityViewer/settings.py [hunk 30]
```diff
@@ -0,0 +1,136 @@
...
+
+# SECURITY WARNING: don't run with debug turned on in production!
+DEBUG = True
+
+ALLOWED_HOSTS = ["localhost", "127.0.0.1", "0.0.0.0"]
+
+# config.DATABASE_URL = "bolt://neo4j:password@localhost:7687"
+config.DATABASE_URL = "bolt://neo4j:password@neo4j_db:7687"
```
Comment (JasperCraeghs) [inline traceabilityViewer/traceabilityViewer/settings.py]: This should become configurable.

File: README.md (hunk 25 above)
Comment (JokeWaumans) [reply README.md]: Suggestion applied in 81ae508d368c73c4aa3a89a1b19e08fe07823e90

File: traceabilityViewer/traceabilityViewer/settings.py [hunk 31]
```diff
@@ -0,0 +1,136 @@
...
+
+
+# Quick-start development settings - unsuitable for production
+# See https://docs.djangoproject.com/en/4.2/howto/deployment/checklist/
+
+# SECURITY WARNING: keep the secret key used in production secret!
+SECRET_KEY = "django-insecure-9@1+y@2v04x9t4r=ih5-%j7ueh(84j6($3tt6js&pd%ozl91du"
+
+# SECURITY WARNING: don't run with debug turned on in production!
+DEBUG = True
```
Comment (JasperCraeghs) [inline traceabilityViewer/traceabilityViewer/settings.py]: Based on that warning, this should become configurable as well.

File: README.md (hunk 12 above)
Comment (JokeWaumans) [reply README.md]: Resolved in bad373c6c3bf6e5eb1e3467b736dfc50a3da6f1e

File: README.md (hunk 13 above)
Comment (JokeWaumans) [reply README.md:44]: Resolved in 388e0b3c04206b709422cca54a34ccaab5d2bd77

File: README.md (hunk 17 above)
Comment (JokeWaumans) [reply README.md]: Resolved in 892a28914e61675a6379ff69b3a4e8c11bd59fed

File: README.md (hunk 19 above)
Comment (JokeWaumans) [reply README.md]: Suggestion applied in commit 9781b6a7c6706a1562ff853f01c4114f28604b40

File: README.md (hunk 20 above)
Comment (JokeWaumans) [reply README.md]: Suggestion applied in commit 7ee06db64ba6d1a14607ea363ebe43ae66775260

File: README.md (hunk 21 above)
Comment (JokeWaumans) [reply README.md]: Suggestion applied in commit 7ee06db64ba6d1a14607ea363ebe43ae66775260

File: README.md (hunk 22 above)
Comment (JokeWaumans) [reply README.md:319]: I mean to use for example modals: https://getbootstrap.com/docs/5.3/components/modal/#static-backdrop
or toasts: https://getbootstrap.com/docs/5.3/components/toasts/#live-example

A modal that needs to be closed to go further of toasts that stay on your screen even when scrolling

File: README.md (hunk 23 above)
Comment (JokeWaumans) [reply README.md]: Suggestion applied in commit 065f4d44857d67b29d971145dcb146c2da52efd9

File: README.md (hunk 29 above)
Comment (JokeWaumans) [reply README.md]: Suggestion applied in commit 815d068e34420389f6cb96f9272f34c59d13ea1d

File: traceabilityViewer/traceabilityViewer/settings.py (hunk 30 above)
Comment (JokeWaumans) [reply traceabilityViewer/traceabilityViewer/settings.py]: Resolved in 0b18671faf1744006b762826255bb4df5a76b926

File: README.md (hunk 16 above)
Comment (JokeWaumans) [reply README.md]: Resolved in d6015a72219f992e6826a558aa3cedc14dd71da2

File: README.md (hunk 14 above)
Comment (JokeWaumans) [reply README.md:65]: Resolved in b676f408445bfd41ceb439da57a63ac4d72f77ac and 11f1ae6b7cbd81beaaadf9b001373c7e1cdd7ffa

File: traceabilityViewer/traceabilityViewer/settings.py (hunk 28 above)
Comment (JokeWaumans) [reply traceabilityViewer/traceabilityViewer/settings.py]: Should I also use `python-decouple` for the other environment variables?
I use for the other variables `os.environ.get("VARIABLE")` because a string was fine.

Comment (Letme) [reply traceabilityViewer/traceabilityViewer/settings.py]: Yes. Usually, `python-decouple` is used so that you can easily switch among environments and it on the end does the same call if it cannot find the variable in the configuration file.

File: traceabilityViewer/traceabilityViewer/settings.py (hunk 31 above)
Comment (JokeWaumans) [reply traceabilityViewer/traceabilityViewer/settings.py:30]: Resolved in ecb55b948db709b70506e6b9c1d0b6602465fa58

Comment (JokeWaumans) [reply traceabilityViewer/traceabilityViewer/settings.py]: Resolved in 8d6e5d4566ca65f4d6965480bd60c60ce5ce4a07

Comment (JasperCraeghs) [reply traceabilityViewer/traceabilityViewer/settings.py]: Better, but you didn't make it configurable. Isn't it useful to have debug turned on during development?

File: README.md [hunk 32]
```diff
@@ -1 +1,309 @@
...
+![error](images/show_error.png)
+
+You can close the error by clicking **X**.
+If the error is not closed and another error occurs, all errors will be listed at the top of the page.
+
+# Bugs
+
+1. When showing the labels/names of the nodes, there is always one node that turns black.
+1. When using the input fields, the autocompletion does not add the selected word at the right place, when adding something in the middle.
+1. When data is obtained via a Cypher query, some links are not visible. When clicking the eye-icon, these nodes that are connected are visible but the links aren't. You can solve it by clicking on the `Add/show connecting nodes` button.
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
1. When data is obtained via a Cypher query, some links are not visible. When clicking on the eye-icon, connected
   nodes are visible but the links are not. You can solve it by clicking on the `Add/show connecting nodes` button.
```
Improved readability and more focus on **not** visible.

File: README.md (hunk 22 above)
Comment (JasperCraeghs) [reply README.md:319]: I like Bootstrap's toasts. It doesn't interrupt the user's workflow compared to scrolling to the top or a modal that blocks progress.

File: README.md [hunk 33]
```diff
@@ -1 +1,309 @@
...
+    - [Errors](#errors)
+- [Bugs](#bugs)
+- [Features](#features)
+
+
+# Overview
+
+The traceability visualizes a collection of traceable items in a network graph. This allows the user to gain a better understanding of the traceability between the different items. The network can be layered or force-directed. This and more options are [configurable](#configuration).
+
+It's originally designed to use the JSON database of our traceability plugin as input but it can also be used with other JSON databases as well. The minimum required JSON database is given in [this section](#json-database).
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
This traceability-viewer, a standalone tool, visualizes a collection of traceable documentation items in a network graph.
This allows the user to gain a better understanding of the relationships between different items.
Each node in the network graph represents an item, and can be grouped in layers, based on its identifier,
or force-directed. This and more options are [configurable](#configuration).

It was originally designed to use the JSON database exported by our
[Sphinx traceability plugin](https://github.com/melexis/sphinx-traceability-extension) as input
but any JSON database can work if it meets the minimum requirements, explained in [this section](#json-database).
```

File: .env.example [hunk 34]
```diff
@@ -0,0 +1,8 @@
+JSON_EXPORT=traceability_export.json
+BASE_URL=https://base_url_to_documentation/or/base_path_to_documentation
+IP_ADDRESS=0.0.0.0
+CONFIG_FILE=config.yml
+PRODUCT=11111
+PACKAGE_TAG=latest
+DATABASE_URL=bolt://neo4j:password@neo4j_db:7687
+SECRET_KEY=ARANDOMSECRETKEY
```
Comment (JasperCraeghs) [inline .env.example:9]: How do all of these environment variables become available in the Docker image/container?

Comment (JokeWaumans) [reply .env.example:9]: via `docker-compose.yml`
```
env_file:
      - .env
```
The ones needed in `Dockerfile` are listed here:
```
build:
      context: .
      args:
        CONFIG_FILE: ${CONFIG_FILE}
        JSON_EXPORT: ${JSON_EXPORT}
```

File: traceabilityViewer/traceabilityViewer/settings.py (hunk 31 above)
Comment (JokeWaumans) [reply traceabilityViewer/traceabilityViewer/settings.py]: Yes, I think the errors gained from it are also interesting. I will make it configurable in `.env` .

Comment (JokeWaumans) [reply traceabilityViewer/traceabilityViewer/settings.py]: Resolved in b9ca41e7d993cef9602617b2af1166553906128b

File: .env.example [hunk 35]
```diff
@@ -0,0 +1,9 @@
...
+DEBUG=true
```
Comment (JasperCraeghs) [inline .env.example]:
```suggestion
DEBUG=False
```

File: README.md [hunk 36]
```diff
@@ -1 +1,315 @@
...
+
+# Getting Started
+
+We use the tool Docker Compose to define and share multi-container applications. This tool makes use of a `docker-compose.yml` file. The `.env` file is meant to store and define environment variables, which are used in the `docker-compose.yml` file and in Python scripts.
+
+First, you need to [install Docker Compose](https://docs.docker.com/compose/install/linux/#install-the-plugin-manually).
+
+<!-- https://docs.docker.com/engine/reference/commandline/compose_up/ -->
+
+After installation, copy the JSON file that contains the data to the project folder and copy the `.env.example` to `.env`. These environment variables requires modification.
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
After installation, copy the JSON database file to this project's folder, copy the `.env.example` to `.env` and customize it.
```

**Review by JasperCraeghs — APPROVED** [review]

File: README.md [hunk 37]
```diff
@@ -302,8 +302,7 @@ If the error is not closed and another error occurs, all errors will be listed a
 
 1. When showing the labels/names of the nodes, there is always one node that turns black.
 1. When using the input fields, the autocompletion does not add the selected word at the right place, when adding something in the middle.
-1. When data is obtained via a Cypher query, some links are not visible. When clicking on the eye-icon, connected
-   nodes are visible but the links are not. You can solve it by clicking on the `Add/show connecting nodes` button.
+1. When data is obtained via a Cypher query, some links are not visible. When clicking on the eye-icon, connected nodes are visible but the links are not. You can solve it by clicking on the `Add/show connecting nodes` button.
```
Comment (JasperCraeghs) [inline README.md]: GitHub rendered it just fine before this change of yours. Why did you change it? Long line lengths are not very nice to review and maintain.

https://github.com/melexis/traceability-viewer/blob/620324f5ca98fd9f79335a618733a09e96083e7e/README.md

image

Comment (JokeWaumans) [reply README.md]: I removed the newline in the middle of the sentence because it would result in this:
image

But I can add newlines after a sentence?

Comment (JokeWaumans) [reply README.md]: I think I resolved it in 30a2aedf809fbcfa11292f6d77f54dfaec8458c7

File: README.md [hunk 38]
```diff
@@ -1 +1,320 @@
...
+
+# Bugs
+
+1. When showing the labels/names of the nodes, there is always one node that turns black.
+1. When using the input fields, the autocompletion does not add the selected word at the right place,
+when adding something in the middle.
+1. When data is obtained via a Cypher query, some links are not visible.
+When clicking on the eye-icon, connected nodes are visible but the links are not.
+You can solve it by clicking on the `Add/show connecting nodes` button.
+1. Hiding a group after selecting a node from that same group, the info of that node is still there.
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
1. When hiding a group while a node of that group was selected, the info of that node remains visible.
```

File: README.md [hunk 39]
```diff
@@ -1 +1,320 @@
...
+### Highlight Connected Nodes
+
+The button with the eye-icon highlights all nodes that are directly connected to the selected node and makes other nodes transparent.
+
+![highlight connected nodes](images/highlight_connected_nodes.png)
+
+
+### Add/Show Connecting Nodes
+
+When the button **Add/show connecting nodes** is clicked, the nodes connecting to the selected node will be obtained. It will add new nodes or links, if there are any.
```
Comment (JasperCraeghs) [inline README.md]:
```suggestion
When you selected a node and click the button **Add/show connecting nodes**, all nodes that are connected to it are obtained and displayed, if there are any.
```
Let's avoid the passive tense in cases like this. It improves readability.

File: README.md (hunk 37 above)
Comment (JasperCraeghs) [reply README.md]: This syntax seems to work in Markdown. FYI In reStructuredText, you would need to align the text:
```
- Bullets are "-", "*" or "+".
  Continuing text must be aligned
  after the bullet and whitespace
  ```
  https://docutils.sourceforge.io/docs/user/rst/quickref.html#bullet-lists
